File: cpp/D2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <regex>
#include "time.h"

using namespace std;

struct policy {
    int min;
    int max;
    char c;
    string password;
    policy(int min, int max, char c, string password)
     : c(c), min(min), max(max), password(password) {}
};

vector<policy> read_input() {
    ifstream input("inputs\\D2.input");
    vector<policy> policies;
    string line;
    regex rgx("([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)");
    while (getline(input, line)) {
        smatch rgxmatch;
        regex_match(line, rgxmatch, rgx);
        policies.push_back(policy(
            stoi(rgxmatch[1]),
            stoi(rgxmatch[2]),
            string(rgxmatch[3])[0],
            rgxmatch[4]
        ));
    }
    return policies;
}


int P1(vector<policy> policies) {
    int valid = 0;

    for (policy p : policies) {
        int count = 0;
        for (char c : p.password) {
            if (c == p.c) count++;
            if (count > p.max) break;
        }
        if (count >= p.min && count <= p.max) valid++;
    }

    return valid;
}

int P2(vector<policy> policies) {
    int valid = 0;

    for (policy p : policies) {
        if (p.password[p.min-1] == p.c 
          ^ p.password[p.max-1] == p.c)
            valid++;
    }

    return valid;
}

int main()
{
    run_with_time<int>([](void) { 
        return P1(read_input()); 
    });

    run_with_time<int>([](void) {
        return P2(read_input());
    });
}

File: ts/D2.ts
import { readFileSync } from 'fs';
import { timeAndPrint } from './time-and-print';

class policy {
    low: number;
    high: number;
    rule: string;
    password: string;
}

function parseInput(): policy[] {
    const policies: policy[] = [];
    const content = readFileSync('../inputs/D2.input','utf8');
    for (let line of content.split("\r\n")) {
        const match = line.match(/([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)/)
        policies.push(<policy>{
            low: +match[1],
            high: +match[2],
            rule: match[3],
            password: match[4]
        });
    }
    return policies;
}

function p1() {
    const policies = parseInput();
    let valid = 0;

    for (let policy of policies) {
        let count = 0;

        for (let char of policy.password) {
            if (char == policy.rule) count++;
            if (count > policy.high) break;
        }

        if (count >= policy.low && count <= policy.high)
            valid++;
    }

    return valid;
}

function p2() {
    const policies = parseInput();
    let valid = 0;

    for (let policy of policies) {
        if ((policy.password[policy.low - 1] == policy.rule) 
        !== (policy.password[policy.high - 1] == policy.rule))
            valid ++;
    }

    return valid;
}

timeAndPrint(p1);
timeAndPrint(p2);